Userland/Libraries/LibWeb/CSS/StyleValues/CalculatedStyleValue.cpp

```diff
@@ -390,6 +390,317 @@ ErrorOr<void> InvertCalculationNode::dump(StringBuilder& builder, int indent) co
+            return { Angle::make_degrees(value) };
+        case CalculatedStyleValue::ResolvedType::Frequency:
+            return { Frequency::make_hertz(value) };
+        case CalculatedStyleValue::ResolvedType::Length:
+            return { Length::make_px(value) };
+        case CalculatedStyleValue::ResolvedType::Percentage:
+            return { Percentage(value) };
+        case CalculatedStyleValue::ResolvedType::Time:
+            return { Time::make_seconds(value) };
+        }
+
+        VERIFY_NOT_REACHED();
+    };
+
+    double square_sum = 0.0;
+
+    for (auto const& value : m_values) {
+        auto child_resolved = value->resolve(layout_node, percentage_basis);
+        auto child_value = resolve_value(child_resolved.value(), layout_node);
+
+        square_sum += child_value * child_value;
+    }
+
+    auto result = sqrt(square_sum);
+
+    return to_resolved_type(resolved_type().value(), result);
+}
+
+ErrorOr<void> HypotCalculationNode::for_each_child_node(Function<ErrorOr<void>(NonnullOwnPtr<CalculationNode>&)> const& callback)
+{
+    for (auto& value : m_values) {
+        TRY(value->for_each_child_node(callback));
+        TRY(callback(value));
+    }
+
+    return {};
+}
+
+ErrorOr<void> HypotCalculationNode::dump(StringBuilder& builder, int indent) const
+{
+    TRY(builder.try_appendff("{: >{}}HYPOT:\n", "", indent));
+    for (auto const& value : m_values)
+        TRY(value->dump(builder, indent + 2));
+    return {};
+}
+
+ErrorOr<NonnullOwnPtr<LogCalculationNode>> LogCalculationNode::create(NonnullOwnPtr<CalculationNode> a, NonnullOwnPtr<CalculationNode> b)
+{
+    return adopt_nonnull_own_or_enomem(new (nothrow) LogCalculationNode(move(a), move(b)));
+}
+
+LogCalculationNode::LogCalculationNode(NonnullOwnPtr<CalculationNode> a, NonnullOwnPtr<CalculationNode> b)
+    : CalculationNode(Type::Log)
+    , m_a(move(a))
+    , m_b(move(b))
+{
+}
+
+LogCalculationNode::~LogCalculationNode() = default;
+
+ErrorOr<String> LogCalculationNode::to_string() const
+{
+    StringBuilder builder;
+    builder.append("log("sv);
+    builder.append(TRY(m_a->to_string()));
+    builder.append(", "sv);
+    builder.append(TRY(m_b->to_string()));
+    builder.append(")"sv);
+    return builder.to_string();
+}
+
+Optional<CalculatedStyleValue::ResolvedType> LogCalculationNode::resolved_type() const
+{
+    return CalculatedStyleValue::ResolvedType::Number;
+}
+
+CalculatedStyleValue::CalculationResult LogCalculationNode::resolve(Layout::Node const* layout_node, CalculatedStyleValue::PercentageBasis const& percentage_basis) const
+{
+    auto node_a = m_a->resolve(layout_node, percentage_basis);
+    auto node_a_value = resolve_value(node_a.value(), layout_node);
+
+    auto node_b = m_b->resolve(layout_node, percentage_basis);
+    auto node_b_value = resolve_value(node_b.value(), layout_node);
+
+    auto result = log2(node_a_value) / log2(node_b_value);
+
+    return { Number(Number::Type::Number, result) };
```

Comment:
it seems that `b`, the base, is optional and equal to _e_ if not given (maybe thereʼs a `ln` function to use instead in such case?)

Reply:
This is already handled during parsing.